### PR TITLE
Move powershell to .NET Core 2.0

### PR DESCRIPTION
**Pull request by daxian-dbw**
This change moves powershell to .NET Core 2.0. Major changes are:
1. PowerShell assemblies are now targeting `netcoreapp2.0`. We are using `microsoft.netcore.app-2.0.0-preview1-001913-00`, which is from dotnet-core build 4/4/17. We cannot target `netstandard2.0` because the packages `System.Reflection.Emit` and `System.Reflection.Emit.Lightweight`, which are needed for powershell class, cannot be referenced when targeting `netstandard2.0`.
2. Refactor code to remove most CLR stub types and extension types.
3. Update build scripts to enable CI builds. ~~Note that the `-cache` option is removed from `appveyor.yml`, but it probably should be added back after this PR is merged.~~ The `-cache` section is added back and specified to depend on `appveyor.yml`, so the cache will be invalidated if `appveyor.yml` is changed. Thanks to @TravisEz13 
4. Ship `netcoreapp` reference assemblies with powershell to fix the issues in `Add-Type` (#2764). By default `Add-Type` will reference all those reference assemblies when compiling C# code. If `-ReferenceAssembly` is specified, then we search reference assemblies first, then the framework runtime assemblies, and lastly the loaded assemblies (possibly a third-party one that was already loaded).
5. `dotnet publish` generates executable on Unix platforms, but doesn't set "x" permission and thus it cannot execute. Currently, the "x" permission is set in the build script, `dotnet/cli` issue [#6286](https://github.com/dotnet/cli/issues/6286) is tracking this.
6. Replace the use of some APIs with the ones that take `SecureString`.
7. osx.10.12 is required to update to `netcoreapp2.0` because `dotnet-cli` 2.0.0-preview only works on osx.10.12.
8. Add dependency to `System.ValueTuple` to work around a ambiguous type identity issue in coreclr. The issue is tracked by `dotnet/corefx` [#17797](https://github.com/dotnet/corefx/issues/17797). When moving to newer version of `netcoreapp2.0`, we need to verify if this dependency is still needed.

### CODE REVIEW

File: appveyor.yml [hunk 1]
```diff
@@ -2,10 +2,6 @@ version: 6.0.0-alpha.18-{build}
 
 image: Visual Studio 2015
 
-cache:
-  - '%LocalAppData%\Microsoft\dotnet'
-  - '%HOMEDRIVE%%HOMEPATH%\.nuget\packages'
```
Comment (TravisEz13) [inline appveyor.yml]: We don't want to cache these at all?

Comment (lzybkr) [inline appveyor.yml]: Can't we clear the cache with a REST call instead of this change?

Comment (TravisEz13) [inline appveyor.yml]: There is also a language to describe when to clear the cache https://www.appveyor.com/docs/build-cache/#cache-dependencies

Comment (daxian-dbw) [inline appveyor.yml]: As I mentioned in the description:
> Note that the `-cache` option is removed from appveyor.yml, but it probably should be added back after this PR is merged.

I planned to added it back after this PR gets merged, and add one more directory path: `%HOMEDRIVE%%HOMEPATH%\.dotnet`. `dotnet-cli` 2.0.0 depends on such a folder now.

Comment (daxian-dbw) [inline appveyor.yml]: @TravisEz13 talked to me offline and it turns out it's quite easy to invalidate the cache, I will do that.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Start-Sleep.Tests.ps1 [hunk 2]
```diff
@@ -1,20 +1,17 @@
...
-        $dtStart = [DateTime]::Now
+    It "Should work properly when sleeping with Second" {
+        $watch = [System.Diagnostics.Stopwatch]::StartNew()
         Start-Sleep -Seconds 1
-        $dtEnd = [DateTime]::Now
-        $milliseconds = (New-TimeSpan -Start $dtStart -End $dtEnd).TotalMilliseconds
-        $milliseconds | Should BeGreaterThan 1000
+        $watch.Stop()
+        $watch.ElapsedMilliseconds | Should BeGreaterThan 950
     }
```
Comment (lzybkr) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Start-Sleep.Tests.ps1]: A comment would be nice - because 1000 seems like a more reasonable value.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Get-Member.Tests.ps1 [hunk 3]
```diff
@@ -256,7 +256,7 @@ Describe "Get-Member DRT Unit Tests" -Tags "CI" {
     Context "Verify Get-Member with other parameters" {
         It 'works with View Parameter' {
             $results = [xml]'<a>some text</a>' | Get-Member -view adapted
-            $results.Length | Should Be 38
```
Comment (lzybkr) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Get-Member.Tests.ps1]: I don't think this is a good or useful test - a better test would sample for some expected members.

File: src/System.Management.Automation/utils/SessionStateExceptions.cs [hunk 4]
```diff
@@ -582,106 +576,6 @@ public SessionStateCategory SessionStateCategory
         #endregion Private
     } // SessionStateException
 
-    // This class is no longer used, but kept in Windows PowerShell for backwards compatibility
```
Comment (lzybkr) [inline src/System.Management.Automation/utils/SessionStateExceptions.cs]: We probably shouldn't delete !CORECLR code in the PR.

File: src/System.Management.Automation/namespaces/SafeRegistryHandle.cs [hunk 5]
```diff
@@ -52,13 +46,9 @@ internal SafeRegistryHandle(IntPtr preexistingHandle, bool ownsHandle) : base(ow
 
         protected override bool ReleaseHandle()
         {
-#if UNIX
-            return true;
```
Comment (lzybkr) [inline src/System.Management.Automation/namespaces/SafeRegistryHandle.cs]: I thought registry apis would throw on Unix - so why is this changed?

File: src/System.Management.Automation/engine/ExtraAdapter.cs [hunk 6]
```diff
@@ -16,220 +15,6 @@
 namespace System.Management.Automation
 {
     /// <summary>
```
Comment (lzybkr) [inline src/System.Management.Automation/engine/ExtraAdapter.cs]: Why is this deleted? If it's Windows PowerShell only now - it should be a different PR.

File: src/System.Management.Automation/engine/CoreAdapter.cs [hunk 7]
```diff
@@ -4661,6 +4662,221 @@ protected override PSMemberInfoInternalCollection<T> GetMembers<T>(object obj)
     }
 
     /// <summary>
+    /// Deals with DataRow objects
```
Comment (lzybkr) [inline src/System.Management.Automation/engine/CoreAdapter.cs]: Oh, I see, you just moved this type, but it's hard to see that for sure. Can it stay in the same place to make it easier to see if anything changed - and if it should move, do that in another refactoring PR?

File: src/System.Management.Automation/CoreCLR/CorePsPlatform.cs [hunk 8]
```diff
@@ -410,6 +410,73 @@ public static string SelectProductNameForDirectory(Platform.XDG_Type dirpath)
...
+
+        /// <summary>
+        /// The API set 'api-ms-win-shell-shellfolders-l1-1-0.dll' was removed from NanoServer, so we cannot depend on 'SHGetFolderPathW'
+        /// to get the special folder paths. Instead, we need to rely on the basic environment variables to get the special folder paths.
+        /// </summary>
+        /// <returns>
+        /// The path to the specified system special folder, if that folder physically exists on your computer.
+        /// Otherwise, an empty string ("").
+        /// </returns>
+        private static string InternalGetFolderPath(System.Environment.SpecialFolder folder)
```
Comment (lzybkr) [inline src/System.Management.Automation/CoreCLR/CorePsPlatform.cs]: So this code just moved files too? Again, it's hard to tell, and reviewing this for the first time, some of it is weird, like ProgramFiles is /bin, but ProgramFilesX86 is /sur/bin. Weird.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs [hunk 9]
```diff
@@ -1015,103 +1015,145 @@ private void CheckTypesForDuplicates(Assembly assembly)
             }
         }
 
-        private static string s_frameworkFolder = System.IO.Path.GetDirectoryName(typeof(object).GetTypeInfo().Assembly.Location);
+        // We now ship the NetCoreApp2.0 reference assemblies with PowerShell Core, so that Add-Type can work
+        // in a predicted way and won't be broken when we move to newer version of .NET Core.
```
Comment (lzybkr) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs]: >predicted [](start = 16, length = 9)

predictable

**Comment by lzybkr** [comment]
@vors - can you review AddType.cs?

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Start-Sleep.Tests.ps1 (hunk 2 above)
Comment (daxian-dbw) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Start-Sleep.Tests.ps1]: Comment added.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Get-Member.Tests.ps1 (hunk 3 above)
Comment (daxian-dbw) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Get-Member.Tests.ps1]: Change the test to sample for 3 members: `a`, `CreateNode`, `CreateElement`.

File: src/System.Management.Automation/utils/SessionStateExceptions.cs (hunk 4 above)
Comment (daxian-dbw) [inline src/System.Management.Automation/utils/SessionStateExceptions.cs]: This class is added back.

File: src/System.Management.Automation/namespaces/SafeRegistryHandle.cs (hunk 5 above)
Comment (daxian-dbw) [inline src/System.Management.Automation/namespaces/SafeRegistryHandle.cs]: `SafeRegistryHandle` is used for transactioned registry, which is not enabled in powershell core. Actually the file `SafeRegistryHandle.cs` should be excluded from building powershell core.

File: src/System.Management.Automation/engine/CoreAdapter.cs (hunk 7 above)
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/CoreAdapter.cs]: Can I make this an exception? DataRow/DataRowView adapters were enabled in `MshObject.cs` when I was working on the first cut of refactoring to enable the build. `ExtraAdapter.cs` contains the adapters that are not supported in powershell core and is excluded from the build, that's why I moved those 2 adapters to `CoreAdatper.cs`.

File: src/System.Management.Automation/CoreCLR/CorePsPlatform.cs (hunk 8 above)
Comment (daxian-dbw) [inline src/System.Management.Automation/CoreCLR/CorePsPlatform.cs]: This code is moved from `System.Management.Automation.Environment` to `SMA.Platform` because the former class is completely removed because all environment members we extended now come back in `System.Environment`. Maybe we need to open a bug to review all if/defs and Unix specific code?

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs (hunk 9 above)
Comment (daxian-dbw) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs]: Fixed.

File: src/System.Management.Automation/engine/CoreAdapter.cs (hunk 7 above)
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/CoreAdapter.cs]: Never mind, I will revert this change, as I didn't add any tests for the DataRow/DataRowView adapters anyway.

**Comment by daxian-dbw** [comment]
@TravisEz13 @lzybkr @vors comments have been addressed, could you please continue the review? Thanks!

**Review by vors — APPROVED** [review]
This is soooo good! 🎉 

I'm signing up mostly on the `Add-Type` changes with minor comments.

File: appveyor.yml [hunk 10]
```diff
@@ -2,9 +2,10 @@ version: 6.0.0-alpha.18-{build}
 
 image: Visual Studio 2015
 
+# cache version - netcoreapp.2.0.0-preview1-001913-00
```
Comment (daxian-dbw) [inline appveyor.yml]: `005724` is the build of `dotnet-cli` we are currently using, but the `netcoreapp` version is actually [`2.0.0-preview1-001913-00`](https://dotnet.myget.org/feed/dotnet-core/package/nuget/Microsoft.NETCore.App/2.0.0-preview1-001913-00).

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs [hunk 11]
```diff
@@ -1015,103 +1015,145 @@ private void CheckTypesForDuplicates(Assembly assembly)
...
-        private static PortableExecutableReference s_secureStringAssemblyReference =
-            MetadataReference.CreateFromFile(typeof(System.Security.SecureString).GetTypeInfo().Assembly.Location);
+        private bool InMemory { get { return String.IsNullOrEmpty(outputAssembly); } }
 
+        /// <summary>
+        /// Initialize the list of reference assemblies that will be used when '-ReferencedAssemblies' is not specified.
+        /// </summary>
+        private static PortableExecutableReference[] InitDefaultRefAssemblies()
+        {
+            var defaultRefAssemblies = new List<PortableExecutableReference>(150);
```
Comment (daxian-dbw) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs]: Currently, `netcoreapp2.0` comes with 137 reference assemblies, and that's why I choose `150` as the default value. I will add a comment.
By default, the capacity of List is 0, and it increases the size as you add more items. So if the size is roughly known, it's better to specify.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs [hunk 12]
```diff
@@ -1015,103 +1015,145 @@ private void CheckTypesForDuplicates(Assembly assembly)
...
-                if (result != null)
+                // It could be a short assembly name or a full assembly name, but we
+                // alwasy want the short name to find the corresponding assembly file.
+                var assemblyName = new AssemblyName(assembly);
+                refAssemblyDll = assemblyName.Name + ".dll";
+            }
+
+            // We look up in reference/framework only when it's for resolving reference assemblies.
+            // In case of 'Add-Type -AssemblyName' scenario, we don't attempt to resolve against framework assemblies because
+            //   2. Explicitly loading a framework assembly usually is not necessary in PowerShell Core.
```
Comment (daxian-dbw) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs]: Goo catch. Will fix.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs [hunk 13]
```diff
@@ -1195,17 +1237,25 @@ private void CompileSourceToAssembly(string source)
...
-                tempReferences.AddRange(s_autoReferencedAssemblies);
+                var tempReferences = new List<PortableExecutableReference>(s_autoReferencedAssemblies.Value);
+                foreach (string assembly in ReferencedAssemblies)
+                {
+                    if (string.IsNullOrWhiteSpace(assembly)) { continue; }
+                    string resolvedAssemblyPath = ResolveAssemblyName(assembly, true);
 
+                    // Ignore some specified reference assemblies
+                    string fileName = PathType.GetFileName(resolvedAssemblyPath);
+                    if (s_refAssemblyNamesToIgnore.Value.Contains(fileName)) { continue; }
```
Comment (daxian-dbw) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs]: Will add it.

File: src/Microsoft.PowerShell.SDK/Microsoft.PowerShell.SDK.csproj [hunk 14]
```diff
@@ -17,87 +16,18 @@
...
+    <PackageReference Include="System.IO.Packaging" Version="4.4.0-preview1-25204-02" />
+    <PackageReference Include="System.Net.Http.WinHttpHandler" Version="4.4.0-preview1-25204-02" />
+    <PackageReference Include="System.ServiceModel.Duplex" Version="4.4.0-beta-25205-01" />
+    <PackageReference Include="System.ServiceModel.Http" Version="4.4.0-beta-25205-01" />
+    <PackageReference Include="System.ServiceModel.NetTcp" Version="4.4.0-beta-25205-01" />
+    <PackageReference Include="System.ServiceModel.Primitives" Version="4.4.0-beta-25205-01" />
+    <PackageReference Include="System.ServiceModel.Security" Version="4.4.0-beta-25205-01" />
+    <PackageReference Include="System.Text.Encodings.Web" Version="4.4.0-preview1-25204-02" />
+    <PackageReference Include="System.Threading.AccessControl" Version="4.4.0-preview1-25204-02" />
+    <PackageReference Include="System.Private.ServiceModel" Version="4.4.0-beta-25205-01" />
```
Comment (daxian-dbw) [inline src/Microsoft.PowerShell.SDK/Microsoft.PowerShell.SDK.csproj]: `netcoreapp2.0` now includes most of those dependencies by default, and that's why many get removed from SDK.csproj. Now I'm inclined to use `SDK.csproj` as:
1. meta project that references other projects
2. meta project that references those packages that are not directly depended by powershell projects.

**Comment by lzybkr** [comment]
:shipit:

**Review by lzybkr — APPROVED** [review]

**Comment by daxian-dbw** [comment]
@TravisEz13 @vors @lzybkr Thanks for the review!

**Comment by iSazonov** [comment]
@daxian-dbw Great work! Thanks!
